src/x86_64/kernel_machine.h

```diff
@@ -111,13 +128,214 @@ extern void write_xmsr(u64, u64);
+}
+
+static inline void cpu_setgs(int cpu)
+{
+    u64 addr = u64_from_pointer(cpuinfo_from_id(cpu));
+    write_msr(KERNEL_GS_MSR, 0); /* clear user GS */
+    write_msr(GS_MSR, addr);
+}
+
+static inline cpuinfo current_cpu(void)
+{
+    u64 addr;
+    asm volatile("movq %%gs:0, %0":"=r"(addr));
+    return (cpuinfo)pointer_from_u64(addr);
+}
+
+static inline u64 extended_frame_size(void)
+{
+#if 0
+    u32 v[4];
+    cpuid(0xd, 0, v);
+    return v[1];
+#else
+    return 512;                 /* XXX fx only right now */
+#endif
+}
+
+static inline u64 total_frame_size(void)
+{
+    return FRAME_EXTENDED_SAVE * sizeof(u64) + extended_frame_size();
+}
+
+/* stubs, for intel sdm recommends using xsave* over manual lazy save/restore */
+#define frame_save_fpsimd(f)
+#define frame_restore_fpsimd(f)
+
+static inline void frame_enable_interrupts(context f)
+{
+    f[FRAME_FLAGS] |= U64_FROM_BIT(FLAG_INTERRUPT);
+}
+
+extern void xsave(context f);
+extern void clone_frame_pstate(context dest, context src);
+extern void init_frame(context f);
+
+static inline boolean is_protection_fault(context f)
+{
+    return (f[FRAME_ERROR_CODE] & FRAME_ERROR_PF_P) != 0;
+}
+
+static inline boolean is_usermode_fault(context f)
+{
+    return (f[FRAME_ERROR_CODE] & FRAME_ERROR_PF_US) != 0;
+}
+
+static inline boolean is_instruction_fault(context f)
+{
+    return (f[FRAME_ERROR_CODE] & FRAME_ERROR_PF_ID) != 0;
+}
+
+static inline boolean is_data_fault(context f)
+{
+    return !is_instruction_fault(f);
+}
+
+static inline boolean is_write_fault(context f)
+{
+    return (f[FRAME_ERROR_CODE] & FRAME_ERROR_PF_RW) != 0;
+}
+
+/* page table integrity check? open to interpretation for other archs... */
+static inline boolean is_pte_error(context f)
+{
+    /* XXX check sdm before merging - seems suspicious */
+    return (is_protection_fault(f) && (f[FRAME_ERROR_CODE] & FRAME_ERROR_PF_RSV));
+}
+
+static inline u64 frame_return_address(context f)
+{
+    return f[FRAME_RIP];
+}
+
+static inline u64 fault_address(context f)
+{
+    return f[FRAME_CR2];
+}
+
+static inline boolean is_page_fault(context f)
+{
+    return f[FRAME_VECTOR] == 14; // XXX defined somewhere?
+}
+
+static inline boolean is_div_by_zero(context f)
+{
+    return f[FRAME_VECTOR] == 0; // XXX defined somewhere?
+}
+
+static inline void frame_save_tls(context f)
+{
+    /* XXX ignore until lazy fs/gs base restore... */
+}
+
+static inline void frame_restore_tls(context f)
+{
+    /* XXX ignore until lazy fs/gs base restore... */
+}
+
+static inline void frame_set_sp(context f, u64 sp)
+{
+    f[FRAME_RSP] = sp;
+}
+
 #define switch_stack(__s, __target) {                           \
         asm volatile("mov %0, %%rdx": :"r"(__s):"%rdx");        \
         asm volatile("mov %0, %%rax": :"r"(__target));          \
         asm volatile("mov %%rdx, %%rsp"::);                     \
         asm volatile("jmp *%%rax"::);                           \
     }
 
+/* syscall entry */
+
```

Comment:
This is just a little thing, but I think this comment is out of place--looks like it should go with init_syscall_handler further up (as that's how it's used in aarch)